src/model_repository_manager.cc

```diff
@@ -665,35 +697,69 @@ ModelRepositoryManager::LoadUnloadModels(
       }
 
       // Only update the infos when all validation is completed
+      ModelInfoMap current_infos;
+      CopyModelInfos(&current_infos);
       for (const auto& model_name : added) {
         auto nitr = new_infos.find(model_name);
-        infos_.emplace(model_name, std::move(nitr->second));
+        current_infos.emplace(model_name, std::move(nitr->second));
       }
       for (const auto& model_name : modified) {
         auto nitr = new_infos.find(model_name);
-        auto itr = infos_.find(model_name);
+        auto itr = current_infos.find(model_name);
         itr->second = std::move(nitr->second);
       }
+      new_infos.swap(current_infos);
+    }
+
+    // Update dependency graph and load
+    RETURN_IF_ERROR(UpdateDependencyGraph(
+        added, deleted, modified, new_infos, &new_dependency_graph,
+        unload_dependents ? &deleted_dependents : nullptr));
+    UpdateTransition(new_dependency_graph, added, true);
+    UpdateTransition(new_dependency_graph, deleted, true);
+    UpdateTransition(new_dependency_graph, modified, true);
+    UpdateTransition(new_dependency_graph, deleted_dependents, true);
+
+    // The models are in 'deleted' either when they are asked to be unloaded or
+    // they are not found / are duplicated across all model repositories.
+    // In all cases, should unload them and remove from 'infos_' explicitly.
+    for (const auto& name :
+         (unload_dependents ? deleted_dependents : deleted)) {
+      new_infos.erase(name);
     }
-  }
-  std::set<std::string> deleted_dependents;
 
-  // Update dependency graph and load
-  UpdateDependencyGraph(
-      added, deleted, modified,
-      unload_dependents ? &deleted_dependents : nullptr);
+    // Write intermediate state
+    infos_.swap(new_infos);
+    dependency_graph_.swap(new_dependency_graph);
+    CopyModelInfos(&new_infos);
+    CopyDependencyGraph(&new_dependency_graph);
+  }
 
-  // The models are in 'deleted' either when they are asked to be unloaded or
-  // they are not found / are duplicated across all model repositories.
-  // In all cases, should unload them and remove from 'infos_' explicitly.
   for (const auto& name : (unload_dependents ? deleted_dependents : deleted)) {
-    infos_.erase(name);
     model_life_cycle_->AsyncUnload(name);
   }
-
+  // do not mark the current affected models as in transition, to distinguish
+  // between models loading/unloading by other threads
+  UpdateTransition(new_dependency_graph, added, false);
+  UpdateTransition(new_dependency_graph, deleted, false);
+  UpdateTransition(new_dependency_graph, modified, false);
+  UpdateTransition(new_dependency_graph, deleted_dependents, false);
```

Comment:
Okay for now, you can consider provide some "set union" function with variable argument lists to merge the sets into single set. i.e. `JoinPath()`

Reply:
👌